runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.http.interceptors
+
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.client.ProtocolResponseInterceptorContext
+import aws.smithy.kotlin.runtime.http.operation.HttpOperationContext
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.http.response.header
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import aws.smithy.kotlin.runtime.time.Instant
+import kotlin.coroutines.coroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+
+/**
+ * An interceptor used to detect clock skew (difference between client and server clocks) and apply a correction.
+ */
+public class ClockSkewInterceptor : HttpInterceptor {
+    public companion object {
+        /**
+         * How much must the clock be skewed before attempting correction
+         */
+        public val CLOCK_SKEW_THRESHOLD: Duration = 4.minutes
+
+        /**
+         * Get the current skew between the client time and [serverTime] as a [Duration].
+         * It may be negative if the serverTime is in the past.
+         * @param serverTime the server's time
+         */
+        public fun Instant.getSkew(serverTime: Instant): Duration = this.until(serverTime)
+
+        /**
+         * Determine whether the client's clock is skewed relative to the server.
+         * @param serverTime the server's time
+         */
+        public fun Instant.isSkewed(serverTime: Instant): Boolean = getSkew(serverTime).absoluteValue >= CLOCK_SKEW_THRESHOLD
+    }
+
+    // Clock skew to be applied to all requests
+    private var currentSkew: Duration = Duration.ZERO
+
+    /**
+     * Apply the previously-computed skew, if it's set, to the execution context before signing
+     */
+    public override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        if (currentSkew != Duration.ZERO) {
+            logger.info { "applying clock skew $currentSkew to client" }
+            context.executionContext[HttpOperationContext.ClockSkew] = currentSkew
+        }
+
+        return context.protocolRequest
+    }
+
+    /**
+     * After receiving a response, check if the client clock is skewed and apply a correction if necessary.
+     */
+    public override suspend fun modifyBeforeDeserialization(context: ProtocolResponseInterceptorContext<Any, HttpRequest, HttpResponse>): HttpResponse {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val clientTime = context.protocolRequest.headers["x-amz-date"]?.let {
```

Comment:
**question**: Should we also be looking at `Date` header on the request and take first one that we find out of (`x-amz-date`, `Date`)?

Also should this check come after looking for server time and fallback to `Instant.now()`? That could be risky if the delta between when the server set the response header and when we get to here is large but in practice it should be way less than the skew threshold so it seems reasonable safe as fallback. This appears to be what Java [does](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/retry/ClockSkewAdjuster.java#L50).

Reply:
From my tests, we don't send a `Date` header, but sure, I can check both headers and then fallback to `Instant.now()`